crates/durability/src/imp/local.rs

```diff
@@ -0,0 +1,303 @@
+            queue_depth,
+            persister_task,
+        })
+    }
+
+    /// Inspect how many transactions added via [`Self::append_tx`] are pending
+    /// to be applied to the underlying [`Commitlog`].
+    pub fn queue_depth(&self) -> u64 {
+        self.queue_depth.load(Relaxed)
+    }
+
+    /// Obtain an iterator over the [`Commit`]s in the underlying log.
+    pub fn commits_from(&self, offset: TxOffset) -> impl Iterator<Item = Result<Commit, error::Traversal>> {
+        self.clog.commits_from(offset)
+    }
+
+    /// Apply all outstanding transactions to the [`Commitlog`] and flush it
+    /// to disk.
+    ///
+    /// Returns the durable [`TxOffset`], if any.
+    pub async fn close(self) -> anyhow::Result<Option<TxOffset>> {
+        info!("close local durability");
+
+        drop(self.queue);
+        if let Err(e) = self.persister_task.await {
+            if e.is_panic() {
+                return Err(e).context("persister task panicked");
+            }
+        }
+
+        spawn_blocking(move || self.clog.flush_and_sync())
+            .await?
+            .context("failed to sync commitlog")
+    }
+}
+
+struct PersisterTask<T> {
+    clog: Arc<Commitlog<Txdata<T>>>,
+    rx: mpsc::UnboundedReceiver<Txdata<T>>,
+    queue_depth: Arc<AtomicU64>,
+    max_records_in_commit: NonZeroU16,
+}
+
+impl<T: Encode + Send + Sync + 'static> PersisterTask<T> {
+    #[instrument(name = "durability::local::persister_task", skip_all)]
+    async fn run(mut self) {
+        info!("starting persister task");
+
+        while let Some(txdata) = self.rx.recv().await {
+            self.queue_depth.fetch_sub(1, Relaxed);
+            trace!("received txdata");
+
+            // If we are writing one commit per tx, trying to buffer is
+            // fairly pointless. Immediately flush instead.
+            //
+            // Otherwise, try `Commitlog::append` as a fast-path which doesn't
+            // require `spawn_blocking`.
+            if self.max_records_in_commit.get() == 1 {
+                self.flush_append(txdata, true).await;
+            } else if let Err(retry) = self.clog.append(txdata) {
+                self.flush_append(retry, false).await
+            }
+
+            trace!("appended txdata");
+        }
+
+        info!("exiting persister task");
+    }
+
+    #[instrument(skip_all)]
+    async fn flush_append(&self, txdata: Txdata<T>, flush_after: bool) {
+        let clog = self.clog.clone();
+        let task = spawn_blocking(move || {
+            let mut retry = Some(txdata);
+            while let Some(txdata) = retry.take() {
+                if let Err(error::Append { txdata, source }) = clog.append_maybe_flush(txdata) {
+                    flush_error(source);
+                    retry = Some(txdata);
+                }
+            }
+
+            if flush_after {
+                clog.flush().map(drop).unwrap_or_else(flush_error);
+            }
+
+            debug!("flush-append succeeded");
```

Comment:
Kurtis asks that we drop this to `trace`.

Reply:
I would have preferred to update the log.conf -- the distinction between debug and trace is like between 10pt and 8pt.

But well...